### PR TITLE
Fixes #3821 copy constructor by making the template molecule a shared…

### LINKED ISSUE #3821: Tautomer Query copy constructor is shallow not deep causing segfaults in destructor
Failing Test (for tautomerQueryTestCatch)

```
TEST_CASE("TAUTOMERQUERY_COPY_CONSTRUCTOR") {
  auto mol = "c1ccccc1"_smiles;
  auto tautomerQuery =
    std::unique_ptr<TautomerQuery>(TautomerQuery::fromMol(*mol));
  auto tautomerQueryCopyConstructed = std::unique_ptr<TautomerQuery>(new TautomerQuery(*tautomerQuery));
}
```

### PR DESCRIPTION
**Pull request by bp-kelley**
Fixes copy constructor for TautomerQueries

### CODE REVIEW

**Comment by bp-kelley** [comment]
@greglandrum @jones-gareth Here is the proposed fix for #3821

**Comment by bp-kelley** [comment]
@greglandrum @jones-gareth copy constructor now does a deep copy.  I had to change the const RO_SPTR vector and switched to a unique_ptr for the template molecule.

**Comment by jones-gareth** [comment]
Looks good to me.

File: Code/GraphMol/TautomerQuery/TautomerQuery.cpp [hunk 1]
```diff
@@ -243,7 +241,7 @@ std::vector<MatchVectType> TautomerQuery::substructOf(
   templateParams.extraFinalCheck = checker;
 
   auto matches =
-      RDKit::SubstructMatch(mol, *d_templateMolecule, templateParams);
+      RDKit::SubstructMatch(mol, *d_templateMolecule.get(), templateParams);
```
Comment (greglandrum) [inline Code/GraphMol/TautomerQuery/TautomerQuery.cpp:244]: I don't think you need the `.get()` here:
```suggestion
      RDKit::SubstructMatch(mol, *d_templateMolecule, templateParams);

```

File: Code/GraphMol/TautomerQuery/TautomerQuery.cpp [hunk 2]
```diff
@@ -269,7 +267,7 @@ bool TautomerQuery::isSubstructOf(const ROMol &mol,
 
 ExplicitBitVect *TautomerQuery::patternFingerprintTemplate(
     unsigned int fpSize) {
-  return PatternFingerprintMol(*d_templateMolecule, fpSize, nullptr, nullptr,
+  return PatternFingerprintMol(*d_templateMolecule.get(), fpSize, nullptr, nullptr,
```
Comment (greglandrum) [inline Code/GraphMol/TautomerQuery/TautomerQuery.cpp:270]:
```suggestion
  return PatternFingerprintMol(*d_templateMolecule, fpSize, nullptr, nullptr,

```

File: Code/GraphMol/TautomerQuery/catch_tests.cpp [hunk 3]
```diff
@@ -308,3 +308,10 @@ TEST_CASE("TEST_NOT_TAUTOMER") {
   REQUIRE(target);
   CHECK(tautomerQuery->isSubstructOf(*target));
 }
+
+TEST_CASE("TAUTOMERQUERY_COPY_CONSTRUCTOR") {
```
Comment (greglandrum) [inline Code/GraphMol/TautomerQuery/catch_tests.cpp:312]: I think it's helpful to be able to relate issues directly to the tests for them:
```suggestion
TEST_CASE("github #3821 TAUTOMERQUERY_COPY_CONSTRUCTOR") {
```

**Review by greglandrum — CHANGES_REQUESTED** [review]
Minor cleanup suggestions.

**Comment by bp-kelley** [comment]
@jones-gareth @greglandrum I'll push the changes.  I also had the thought that if I used a shared_ptr instead of a unique_ptr we could have operator= do a shallow copy, as it is, I think it will fail to compile.  If we want this behavior I'll make the change.

**Comment by greglandrum** [comment]
> @jones-gareth @greglandrum I'll push the changes. I also had the thought that if I used a shared_ptr instead of a unique_ptr we could have operator= do a shallow copy, as it is, I think it will fail to compile. If we want this behavior I'll make the change.

What change is this in reference to?

**Comment by bp-kelley** [comment]
@greglandrum @jones-gareth  I was wrong operator= will use the copy constructor, I added a test for this so we're good to go.

**Comment by greglandrum** [comment]
> @greglandrum it's the .get() removal.
> 
> The other piece was if we wanted to support operator= as a shallow copy and the copy constructor as a deep copy. As is operator= won't work with the unique_ptr. So I may make it private.

I still don't get it. I believe that `*d_templateMolecule.get()` is exactly equivalent to `*d_templateMolecule`.

**Comment by bp-kelley** [comment]
@greglandrum it is, looks like I only pushed the header file changes.

**Comment by bp-kelley** [comment]
@greglandrum pushed the .cpp changes.

**Review by greglandrum — APPROVED** [review]

**Pull request merged**